Add Bulletin Chain data storage reference and tutorial (#1575)
polkadot-developers/polkadot-docs: reference/polkadot-hub/data-storage.md

---
title: Data Storage
description: Technical reference for the Polkadot Bulletin Chain, a specialized data storage chain with IPFS-compatible content addressing for the Polkadot ecosystem.
categories: Polkadot Protocol
---

# Data Storage

> The Polkadot Bulletin Chain is for storing raw data and files on-chain — images, HTML, JSON, documents, and more. You get back a CID (Content Identifier) and the data is directly retrievable. Think of it as decentralized file hosting built on a Polkadot SDK-based blockchain with [IPFS](https://ipfs.tech/){target=\_blank}-compatible content addressing. The data has a retention period and needs renewal. There are no ownership semantics — it's just storage.

Reviewer: ```suggestion
The Polkadot Bulletin Chain is for storing raw data and files on-chain — images, HTML, JSON, documents, and more. You get back a CID (Content Identifier) and the data is directly retrievable. Think of it as decentralized file hosting built on a Polkadot SDK-based blockchain with [IPFS](https://ipfs.tech/){target=\_blank}-compatible content addressing. The data is only kept for a specific retention period and needs to be renewed to be kept for longer. There are no ownership semantics — it's just storage.
```

Author: Applied — updated to "The data is only kept for a specific retention period and needs to be renewed to be kept for longer."

> The Bulletin Chain is designed as a utility chain — it has **no native token balances**. Instead, access is managed through an authorization system that grants accounts or preimages the ability to store data. For more details, see the [Polkadot Bulletin Chain repository](https://github.com/paritytech/polkadot-bulletin-chain){target=\_blank}.

Reviewer: We'll soon move it to the fellowship runtimes repo

Author: Noted — added a parenthetical mentioning the upcoming move to the fellowship runtimes repository with a link.

- **Static sites**: Host a static website entirely on-chain, served via IPFS-compatible gateways.
- **Images and media assets**: Store images, icons, or other media files with permanent, verifiable CIDs.
- **Application data**: Persist configuration files, metadata, or other structured data that needs to be publicly accessible and tamper-proof.
- **Document storage**: Store documents, certificates, or records that benefit from on-chain verifiability.

## Key Concepts

- **No balances**: The chain does not use token balances for transaction fees. Authorization is required to store data.
- **Authorization-based access**: Accounts must be authorized before they can submit storage transactions.
- **Content Identifiers (CIDs)**: Stored data is addressable via IPFS-compatible CIDs, enabling interoperability with IPFS tooling.
- **Retention periods**: Stored data is retained for a limited period (~2 weeks on Polkadot TestNet) and must be renewed to persist beyond that.
- **Chunked uploads**: Large files (up to ~64 MiB) can be split into chunks, each stored as a separate transaction (max ~8 MiB per transaction).

## Authorization

> The Bulletin Chain has no token balances — there are no fees to pay. Instead, storage access is controlled through an authorization system that prevents spam and ensures fair usage. A privileged account must explicitly grant permission before an account can store data.

Reviewer: ```suggestion
The Bulletin Chain has no token balances — there are no fees to pay. Instead, storage access is controlled through an authorization system that prevents spam and ensures fair usage. A privileged origin must explicitly grant permission before an account can store data. Currently, only OpenGov can provide authorizations but the PoP (Proof of Personhood) subsystem is also planned to have this privilege in the future.
```

Author: Applied — updated the authorization section to mention OpenGov as the current source and PoP (Proof of Personhood) as a planned future privilege.

Authorization is not a token transfer — it's a **permission entry** recorded in the chain's storage that specifies:

- The **account** allowed to store data
- The number of **transactions** and **bytes** the account can use
- An optional **expiration block** after which the authorization becomes invalid

> Authorization is consumed as you store data — each storage transaction decrements your remaining transaction count and byte allowance. Once your allocation is used up or the expiration block is reached, you need new authorization. Unused authorization is not refunded.

Reviewer: ```suggestion
Authorization is consumed as you store data — each storage transaction decrements your remaining transaction count and byte allowance. Once your allocation is used up or the expiration block is reached, you need a new authorization.
```

This might change in the future to restore the bytes you can use when data is dropped after the retention period

Author: Applied — removed "Unused authorization is not refunded" since this may change when bytes are restored after the retention period.

| `ByteFee` | `Balance` | The fee charged per byte of stored data |
| `EntryFee` | `Balance` | The base fee charged per storage transaction |
| `RetentionPeriod` | `BlockNumber` | The number of blocks for which stored data is retained |

The pallet emits the following events:

| Event | Description |
|:--:|:--:|
| `Stored` | Emitted when data is successfully stored. Contains the transaction index, content hash, and CID |
| `Renewed` | Emitted when a stored transaction is renewed |
| `ProofChecked` | Emitted when a storage proof is successfully verified |
| `AccountAuthorized` | Emitted when an account is granted storage authorization |
| `PreimageAuthorized` | Emitted when a preimage hash is granted storage authorization |
| `AccountAuthorizationRefreshed` | Emitted when an account authorization is refreshed |
| `PreimageAuthorizationRefreshed` | Emitted when a preimage authorization is refreshed |
| `AccountAuthorizationRemoved` | Emitted when an expired account authorization is removed |
| `PreimageAuthorizationRemoved` | Emitted when an expired preimage authorization is removed |

## IPFS Integration and Data Retrieval

> The Bulletin Chain follows a **"write-to-chain, read-from-network"** architecture. Data is stored via on-chain transactions and retrieved directly from validator nodes using the CID. The chain is designed to work seamlessly with the [InterPlanetary File System (IPFS)](https://ipfs.tech/){target=\_blank} — Bulletin Chain nodes are themselves IPFS participants.

Reviewer: ```suggestion
The Bulletin Chain follows a **"write-to-chain, read-from-network"** architecture. Data is stored via on-chain transactions and retrieved directly from collator nodes using the CID. The chain is designed to work seamlessly with the [InterPlanetary File System (IPFS)](https://ipfs.tech/){target=\_blank} — Bulletin Chain nodes are themselves IPFS participants.
```

Author: Applied — updated to collator nodes.

### How It Works

When you store data on the Bulletin Chain, the following happens:

1. **CID generation**: The chain computes an IPFS-compatible [Content Identifier (CID)](https://docs.ipfs.tech/concepts/content-addressing/){target=\_blank} from your data using a cryptographic hash. This CID uniquely identifies the content regardless of where it is stored.

> 2. **Bitswap serving**: Bulletin Chain validator nodes implement the IPFS [Bitswap](https://docs.ipfs.tech/concepts/bitswap/){target=\_blank} wire protocol, which is the standard way IPFS nodes exchange data blocks. IPFS clients can request data directly from validator nodes using the CID.

Reviewer: ```suggestion
2. **Bitswap serving**: Bulletin Chain collator nodes implement the IPFS [Bitswap](https://docs.ipfs.tech/concepts/bitswap/){target=\_blank} wire protocol, which is the standard way IPFS nodes exchange data blocks. IPFS clients can request data directly from these nodes using the CID.
```

Author: Applied — updated to collator nodes.

> 3. **DHT registration**: Validator nodes register as content providers on the IPFS [Kademlia DHT](https://docs.ipfs.tech/concepts/dht/){target=\_blank}, making stored data discoverable by the IPFS network.

Reviewer: I don't think this is correct. We don't want to be part of the larger IPFS network

Author: Removed the DHT registration point entirely, along with the "Content Provider Record Expiry" section that depended on it.

## Authorization

The Bulletin Chain has no token balances — there are no fees to pay. Instead, storage access is controlled through an authorization system that prevents spam and ensures fair usage. A privileged account must explicitly grant permission before an account can store data.

Authorization is not a token transfer — it's a **permission entry** recorded in the chain's storage that specifies:

- The **account** allowed to store data
- The number of **transactions** and **bytes** the account can use
- An optional **expiration block** after which the authorization becomes invalid

Authorization is consumed as you store data — each storage transaction decrements your remaining transaction count and byte allowance. Once your allocation is used up or the expiration block is reached, you need new authorization. Unused authorization is not refunded.

### Who Grants Authorization

The `authorize_account` and `authorize_preimage` extrinsics require **Root origin**, meaning only privileged accounts can grant authorization. How this works depends on the network:

| Network | Authorization Source |

> |:---|:---|

Reviewer 2: ```suggestion
|:--:|:--:|
```
